### PR TITLE
Rc 2

### PR DESCRIPTION
**Pull request by mwalkowski**

### CODE REVIEW

File: src/vmc/assets/documents.py [hunk 1]
```diff
@@ -143,7 +153,11 @@ def get_assets_with_tag(tag: str, config=None):
     def update_gone_discovered_assets(targets, scanned_hosts, discovered_assets, config=None):
         index = AssetDocument.get_index(config)
         # FIXME: update by query
```
Comment (Kraxi) [inline src/vmc/assets/documents.py:155]: should that comment stay?

File: src/vmc/config/settings.py [hunk 2]
```diff
@@ -36,6 +36,18 @@ def get_config(key, default):
     return CFG[key] if CFG and key in CFG else default
 
 
+#VMC absolute URI settings
+
+SCHEMA = 'https' if get_config('vmc.ssl', False) else 'http'
+DOMAIN = get_config('vmc.domain', 'localhost')
+
+if get_config('vmc.port', '80') not in ["443", "80"]:
+    PORT = F":{get_config('vmc.port', '443')}"
```
Comment (Kraxi) [inline src/vmc/config/settings.py:45]: Check why Codecov cries

File: src/vmc/processing/tasks.py [hunk 3]
```diff
@@ -266,9 +267,12 @@ def _processing(idx, slices_count, assets_count, vulnerability_index):
         LOGGER.debug(F'Calculation for {vulnerability_index} and {idx}, {slices_count} started')
 
         if slices_count > 1:
-            vuln_search = vuln_search.extra(slice={"id": idx, "max": slices_count})
+            vuln_search = vuln_search.extra(slice={"id": idx, "max": slices_count}).params(scroll="60m")
 
-        for vuln in vuln_search.scan():
```
Comment (Kraxi) [inline src/vmc/processing/tasks.py:271]: Don't get it

File: src/vmc/assets/documents.py (hunk 1 above)
Comment (mwalkowski) [reply src/vmc/assets/documents.py:155]: Yes, I don't have idea how to make this now

File: src/vmc/processing/tasks.py (hunk 3 above)
Comment (Kraxi) [reply src/vmc/processing/tasks.py:271]: Add a comment to mark this is for performance reasons

File: src/vmc/scanners/nessus/clients.py [hunk 4]
```diff
@@ -121,6 +111,143 @@ def _raise_exception(headers, endpoint, status_code, data=None, content=None):
...
+        super().__init__(config)
+
+    def download_scan(self, scan_id: int, report_format: Client.ReportFormat.XML):
+        extra = self._get_report_format(report_format)
+        res = self._action(F'scans/{scan_id}/export', method="POST", extra=extra)
+
+        if 'file' in res:
+            file_id = res["file"]
+            while self._export_in_progress(scan_id, file_id):
+                time.sleep(2)
```
Comment (Kraxi) [inline src/vmc/scanners/nessus/clients.py:128]: Codecov cries about no test for sleep(2)

File: src/vmc/scanners/nessus/clients.py [hunk 5]
```diff
@@ -121,6 +111,143 @@ def _raise_exception(headers, endpoint, status_code, data=None, content=None):
...
+
+            content = self._action(F'scans/{scan_id}/export/{file_id}/download', method="GET", download=True)
+            return BytesIO(content)
+        return None
```
Comment (Kraxi) [inline src/vmc/scanners/nessus/clients.py:132]: No test for return None

File: src/vmc/scanners/nessus/clients.py [hunk 6]
```diff
@@ -121,6 +111,143 @@ def _raise_exception(headers, endpoint, status_code, data=None, content=None):
...
+    version = r'^8.'
+
+    def download_scan(self, scan_id: int, report_format: Client.ReportFormat.XML):
+        extra = self._get_report_format(report_format)
+        res = self._action(F'scans/{scan_id}/export', method="POST", extra=extra)
+
+        if 'file' in res:
+            token = res["token"]
+            while self._export_in_progress(token):
+                time.sleep(2)
```
Comment (Kraxi) [inline src/vmc/scanners/nessus/clients.py:154]: Same as Nessus7

File: src/vmc/scanners/nessus/clients.py [hunk 7]
```diff
@@ -121,6 +111,143 @@ def _raise_exception(headers, endpoint, status_code, data=None, content=None):
...
+
+            content = self._action(F'tokens/{token}/download', method="GET", download=True)
+            return BytesIO(content)
+        return None
```
Comment (Kraxi) [inline src/vmc/scanners/nessus/clients.py:158]: Same as Nessus7

File: src/vmc/scanners/nessus/clients.py [hunk 8]
```diff
@@ -121,6 +111,143 @@ def _raise_exception(headers, endpoint, status_code, data=None, content=None):
...
+        self._url = config.get_url()
+        self._config = config
+
+        if config.insecure and hasattr(requests, 'packages'):
+            requests.packages.urllib3.disable_warnings()
+
+        try:
+            self._version = self._get_version()
+
+            if re.match(_NessusClient8.version, self._version):
```
Comment (Kraxi) [inline src/vmc/scanners/nessus/clients.py:217]: No test (Codecov)

File: src/vmc/scanners/nessus/clients.py [hunk 9]
```diff
@@ -121,6 +111,143 @@ def _raise_exception(headers, endpoint, status_code, data=None, content=None):
...
+                LOGGER.debug('Using nessus client version 8')
+                self.client = _NessusClient8(config)
+            elif re.match(_NessusClient7.version, self._version):
+                LOGGER.debug('Using nessus client version 7')
+                self.client = _NessusClient7(config)
+            else:
+                raise Exception(F'Unknown nessus version {self._version}')
```
Comment (Kraxi) [inline src/vmc/scanners/nessus/clients.py:224]: No test for exception raising

File: src/vmc/scanners/nessus/clients.py [hunk 10]
```diff
@@ -121,6 +111,143 @@ def _raise_exception(headers, endpoint, status_code, data=None, content=None):
...
+
+        except Exception as ex:
```
Comment (Kraxi) [inline src/vmc/scanners/nessus/clients.py:226]: No test for general exception

File: src/vmc/scanners/nessus/parsers.py [hunk 11]
```diff
@@ -61,29 +62,44 @@ def __init__(self, config: Config):
...
-                    if x['folder_id'] != NessusReportParser.get_trash_folder_id(scan_list)]
+    def get_scans_ids(self, reports) -> List:
+        if reports['scans']:
+            folders = self._get_folders(reports)
+            return [x['id'] for x in reports['scans'] if x['folder_id'] in folders]
         return []
 
-    @staticmethod
-    def get_trash_folder_id(scan_list: Dict) -> [int, None]:
+    def _get_folders(self, scan_list: Dict) -> [int, None]:
```
Comment (Kraxi) [inline src/vmc/scanners/nessus/parsers.py:71]: return type is wrong

File: src/vmc/scanners/openvas/parsers.py [hunk 12]
```diff
@@ -31,24 +32,58 @@
...
+    return cve
+
+
+class GmpParserOMP7(Parser):
 
     def __init__(self, config: Config):
         self._config = config
         self.__parsed = dict()
-        self.__scanned_host = list()
+        self.__scanned_host = {}
```
Comment (Kraxi) [inline src/vmc/scanners/openvas/parsers.py]: dict()

File: src/vmc/scanners/registries.py [hunk 13]
```diff
@@ -22,24 +22,16 @@
 class Registry:
 
     def __init__(self):
-        self.__clients = {}
-        self.__parsers = {}
+        self.__managers = {}
```
Comment (Kraxi) [inline src/vmc/scanners/registries.py]: dict()

File: src/vmc/webhook/thehive/conventer.py [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+                task.document_id = art['data']
+
+        task.title, task.group = self.get_task_title_and_group(alert['title'])
+        task.source = alert['source']
+        task.description = alert['description']
+        task.save()
+
+    @staticmethod
+    def get_task_title_and_group(title):
+        new_title = ''.join(title.split(':')[1:])
```
Comment (Kraxi) [inline src/vmc/webhook/thehive/conventer.py:52]:
```
new_title = ''.join(title.split(':')[1:])
groups = ['critical', 'medium', 'high']

    for g in groups:
        if g in new_title:
            return new_title, g  

    return new_title, 'low'
   ```

File: src/vmc/webhook/thehive/conventer.py [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+    def _build_description(obj):
+        groups = {}
+        for task in obj.task_set.all():
+            if task.group not in groups:
+                groups.update({task.group: 1})
+            else:
+                groups[task.group] += 1
+
+        s = F"""**Summary**:\n
+Critical: {groups['critical'] if 'critical' in groups else 0}\n
```
Comment (Kraxi) [inline src/vmc/webhook/thehive/conventer.py:116]: From this point we should consider moving that to separate file (with ssome constant values or strings) and concatenate

File: src/vmc/webhook/thehive/views.py [hunk 16]
```diff
@@ -0,0 +1,24 @@
+from rest_framework.decorators import api_view
+from rest_framework.response import Response
+
+from vmc.webhook.thehive.handlers import handlers_map
+from vmc.webhook.models import TheHive4
+
+
+def capitalize(event_name):
```
Comment (Kraxi) [inline src/vmc/webhook/thehive/views.py]: should that be treated as string.capitalize() overload??

**Review by Kraxi — COMMENTED** [review]
review the comments and then we'll discuss

**Pull request merged**